### ISSUE TITLE
Increase the expandObject Depth Limit (currently 16)

### ISSUE THREAD

**Issue by Zephyr2011** [foundryvtt/foundryvtt#9711]
### User Experience

Please increase the depth limit of the expandObject function to 32 from 16 (line 257 of helpers.mjs). This will support the use of complex "prefabs" via Token Attacher and their use across multiple systems.

In v10 of Foundry, the depth limit was 100, in v11 it was reduced to 16. This is a great change for preventing recursion issues etc. however, there are a few niche cases where this depth limit greatly affects some modules. Chief among those are "prefabs" created using Token Attacher. Token Attacher uses a conversion script to make prefabs system agnostic by unpacking a JSON of the actors compendium. The depth limit is not normally an issue, however, when there are prefabs that contain advanced pieces such as Monk's Active Tile Triggers enabled tiles, the depth limit is quickly reached resulting in a failure to convert the compendiums.

From testing, increasing the depth limit to 32 is sufficient for even the most advanced prefabs in compendiums to successfully convert. I propose the depth limit be raised to 32 in order to support advanced prefabs and other complex compendium operations while still being substantially shallower than the previous limit of 100.

I'm happy to provide additional information or discuss possible work arounds if needed!

**Comment by aaclayton** [comment]
Hi Zephyr! Could you expand upon the use case a little bit? What do you mean by a "Complex prefab"? Would this be prefab data stored inside a token inside a Scene? If so that is probably an unadvisable data structure to utilize for performance reasons and there may be a better approach to the problem. This safeguard exists - in part - to protect users from harming themselves with unnecessarily complex/recursive/nested data structures. We can consider elevating the limit but I'd like to understand more about the use case and its requirements first.

**Comment by theripper93** [comment]
Since i discussed this issue with Zephyr, i'll chime in to explain the problem:
- Zephyr, among other creators, uses the Token Attacher module to create "prefabs" to distribute modular content
- Due to the fact that it's not possible to create "Token" compendiums, and that actor compendiums are system restricted, Token attacher exports the actor compendiums to a JSON that is then re-imported into different systems using their default actor.
- Token attacher stores placeable documents data into the token flags, when you pair this with other modules, namely Monk's Active Tiles which can have very complex and deep flags data structures, it's pretty easy to exceeed the 16 depth limit by marginal amounts

**Comment by Zephyr2011** [comment]
> Hi Zephyr! Could you expand upon the use case a little bit? What do you mean by a "Complex prefab"? Would this be prefab data stored inside a token inside a Scene? If so that is probably an unadvisable data structure to utilize for performance reasons and there may be a better approach to the problem. This safeguard exists - in part - to protect users from harming themselves with unnecessarily complex/recursive/nested data structures. We can consider elevating the limit but I'd like to understand more about the use case and its requirements first.

Hi Atropos, just to expand upon what Ripper mentioned: "complex prefabs" are token attacher constructs where tiles and other entities are attached to a token, which is then assigned an actor. The complex side of it is when we're attaching tiles and such that have things like Monk's Active Tile Triggers data. This isn't an issue in a scene at all, but we have to export these to actor compendiums, and then export said compendiums to JSON files that Token Attacher uses to create system-agnostic compendiums.

Like Ripper mentioned, the depth issue only occurs when importing those system agnostic prefabs, I *believe* this is because each actor in the compendium has to store all of the data for every piece that's attached to it. Whereas in a scene, it's a lot lighter weight. The token has flags for what's attached to it, and the attached objects have flags for what token they're attached to.

Basically, expanding the depth limit would just allow us to continue using the complex prefabs for non-5e systems. Unfortunately, right now our PF2e and non-5e users won't be able to use our prefabs in v11. If there's other solutions to this, I'm definitely open to it. We tried a few different work arounds in our module script, but kind of ran into a wall without a re-write from KayelGee who's had to step away from development for a bit.

**Comment by Zephyr2011** [comment]
Just following up on this.

Any further questions or considerations here?

**Comment by aaclayton** [comment]
Tagging this for consideration. Will discuss with the team.

I don't personally agree with the technical approach being applied here, but I don't think we need to intentionally block this use case on principle.

**Comment by Gawdl3y** [comment]
I *believe* the reduction to 16 is also affecting the Item Collection module, restricting the depth of containers (container within a container). See its [11.0.0 release notes](https://gitlab.com/tposney/itemcollection/-/blob/v10/Changelog.md#anchor-1100) for the relevant tidbit about this.
